packages/mongo/oplog_tailing.js

```diff
@@ -278,50 +335,25 @@ _.extend(OplogHandle.prototype, {
             continue;
           }
 
-          var doc = self._entryQueue.shift();
+          const doc = self._entryQueue.shift();
 
-          if (!(doc.ns && doc.ns.length > self._dbName.length + 1 &&
-                doc.ns.substr(0, self._dbName.length + 1) ===
-                (self._dbName + '.'))) {
-            throw new Error("Unexpected ns");
-          }
-
-          var trigger = {collection: doc.ns.substr(self._dbName.length + 1),
-                         dropCollection: false,
-                         dropDatabase: false,
-                         op: doc};
-
-          // Is it a special command and the collection name is hidden somewhere
-          // in operator?
-          if (trigger.collection === "$cmd") {
-            if (doc.o.dropDatabase) {
-              delete trigger.collection;
-              trigger.dropDatabase = true;
-            } else if (_.has(doc.o, 'drop')) {
-              trigger.collection = doc.o.drop;
-              trigger.dropCollection = true;
-              trigger.id = null;
-            } else {
-              throw Error("Unknown command " + JSON.stringify(doc));
-            }
-          } else {
-            // All other ops have an id.
-            trigger.id = idForOp(doc);
-          }
-
-          self._crossbar.fire(trigger);
+          // Fire trigger(s) for this doc.
+          handleDoc(doc);
 
           // Now that we've processed this operation, process pending
           // sequencers.
-          if (!doc.ts)
+          if (doc.ts) {
+            self._setLastProcessedTS(doc.ts);
```

Comment:
The `doc.ts` property only exists on the top-level `doc` in case of transactions, which is why I kept this code outside the `handleDoc` function.

Reply:
@benjamn ,

We have been using the `rawCollection` exposed by Meteor in order to implement transactions, and we ran into a bug which we believe is connected to this. In the mongo driver, in `oplog_tailing.js`, we get an error stemming from the `ts` property missing on each object in the `applyOps` array of the transaction oplog entry. 

Here is the stack trace for the error:

```
 Error in oplog callback TypeError: Cannot read property 'toString' of undefined
     at packages/mongo/oplog_observe_driver.js:575:5
     at Object.Meteor._noYieldsAllowed (packages/meteor.js:848:12)
     at OplogObserveDriver._handleOplogEntrySteadyOrFetching (packages/mongo/oplog_observe_driver.js:583:9)
     at packages/mongo/oplog_observe_driver.js:137:3
     at packages/mongo/oplog_observe_driver.js:12:1
     at Object.Meteor._noYieldsAllowed (packages/meteor.js:848:12)
     at packages/mongo/oplog_observe_driver.js:122:13
     at packages/mongo/oplog_tailing.js:105:7
     at runWithEnvironment (packages/meteor.js:1356:24)
     at Object.callback (packages/meteor.js:1369:14)
     at packages/ddp-server/crossbar.js:114:36
     at Array.forEach (<anonymous>)
     at Function._.each._.forEach (packages/underscore.js:139:11)
     at DDPServer._Crossbar.fire (packages/ddp-server/crossbar.js:112:7)
     at handleDoc (packages/mongo/oplog_tailing.js:318:11)
     at Array.forEach (<anonymous>)
```

We tested adding a few lines to fix this, and the error went away:

```javascript
    if (doc.ns === "admin.$cmd") {
          if (doc.o.applyOps) {
            // This was a successful transaction, so we need to apply the
            // operations that were involved.

            // We added this map function to resolve our issue:
            doc.o.applyOps.map(function (op) {
                if (!op.ts) {
                    console.log('applying parent ts')
                    op.ts = doc.ts;
                }
                return op;
            }).forEach(handleDoc);
            return;
          }

          throw new Error("Unknown command " + EJSON.stringify(doc));
        }
```